### PR TITLE
Implementing app settings

### PR DESCRIPTION
**Pull request by MattIPv4**
- Rename About to App Settings
 - Add version information to updates modal
 - Add ability to set MagicCap to open at login
 - Add ability to export all magiccap config options to file
 - Move & update ability to export magiccap uploader config options to file
 - Add ability to export magiccap capture history to file
 - Import config options or history from file
 - Add ability to reset capture history
 - Add ability to reset all settings in MagicCap

### CODE REVIEW

File: src/gui/templates/app_settings.vue [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+                        }
+                        try {
+                            writeFileSync(file, data)
+                        } catch (err) {
+                            console.log(err)
+                        }
+                    })()
+                })
+            },
+            encode(type: String, data: any) {
```
Comment (MattIPv4) [inline src/gui/templates/app_settings.vue]: `String` type is lowercase.

File: src/gui/templates/app_settings.vue [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+                return `==BEGIN MAGICCAP ${type}==\n${btoa(btoa(encodeURIComponent(JSON.stringify(data))))}`
+            },
+            decode(raw: String) {
```
Comment (MattIPv4) [inline src/gui/templates/app_settings.vue]: `String` type is lowercase.

File: src/gui/templates/app_settings.vue [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+                        try {
+                            writeFileSync(file, data)
+                        } catch (err) {
+                            console.log(err)
+                        }
+                    })()
+                })
+            },
+            encode(type: String, data: any) {
+                return `==BEGIN MAGICCAP ${type}==\n${btoa(btoa(encodeURIComponent(JSON.stringify(data))))}`
```
Comment (MattIPv4) [inline src/gui/templates/app_settings.vue]: Split this up to make the conversion clearer.

File: src/gui/templates/app_settings.vue [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+            },
+            decode(raw: String) {
+                return JSON.parse(decodeURIComponent(atob(atob(raw.split("\n").slice(1).join("\n")))))
```
Comment (MattIPv4) [inline src/gui/templates/app_settings.vue]: Split this up to make the conversion clearer.

File: src/gui/templates/app_settings.vue [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+            },
+            exportConfig() {
+                const config = window.config.o
+
+                // Clean install specific items
+                if("ffmpeg_path" in config) delete config.ffmpeg_path
+                if("install_id" in config) delete config.install_id
+
+                // Convert to save format
+                // Double base64 encode to help keep plain-text logins & tokens safer
```
Comment (MattIPv4) [inline src/gui/templates/app_settings.vue]: Move this explanation comment to the encode method.

File: src/gui/templates/app_settings.vue [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+                const data = this.encode("CONFIG", window.config.o)
+
+                // Save
+                this.saveConfig(data)
+            },
+            exportUploaders() {
+                const exported = window.mconf.newConfig()
+
+                // Convert to save format
+                // Double base64 encode to help keep plain-text logins & tokens safer
```
Comment (MattIPv4) [inline src/gui/templates/app_settings.vue]: Move this explanation comment to the encode method.

File: src/gui/templates/app_settings.vue [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+            },
+            exportHistory() {
+                const captures: any[] = []
+                const stmt = db.prepare("SELECT * FROM captures")
+                for (const i of stmt.iterate()) {
+                    captures.push(i)
+                }
+
+                // Convert to save format
+                // Double base64 encode to help keep plain-text logins & tokens safer
```
Comment (MattIPv4) [inline src/gui/templates/app_settings.vue]: Move this explanation comment to the encode method.

File: src/gui/templates/app_settings.vue [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+                const data = this.encode("UPLOADERS", exported)
+
+                // Save
+                this.saveConfig(data)
+            },
+            exportHistory() {
+                const captures: any[] = []
```
Comment (MattIPv4) [inline src/gui/templates/app_settings.vue]: Update type, this is an array of objects.

**Comment by MattIPv4** [comment]
cc @JakeMakesStuff Do we need to set a default loader? Should I just replicate what happens to the config on a new install, but preserve install_id & ffmpeg_path?

**Comment by JakeMakesStuff** [comment]
@MattIPv4 We can just use the default config for that tbh. Thinking about it, we can even re-fetch the install ID if it's easier, they'll always get the same install ID if they've got the same machine ID.

**Pull request merged**